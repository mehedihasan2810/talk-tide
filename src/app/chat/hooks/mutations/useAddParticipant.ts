import { apiClient } from "@/lib/api";
import { ChatInterface } from "@/types/chat";
import { requestHandler } from "@/utils/requestHandler";
import { useMutation } from "@tanstack/react-query";

export const useAddParticipant = () => {
  return useMutation({
    mutationFn: ({
      chatId,
      participantToBeAdded,
    }: {
      chatId: string;
      participantToBeAdded: string;
    }) =>
      requestHandler<ChatInterface>(() =>
        apiClient.post(
          `/chatApp/group/participant/add/${chatId}/${participantToBeAdded}`,
        ),
      )(),
  });
};
