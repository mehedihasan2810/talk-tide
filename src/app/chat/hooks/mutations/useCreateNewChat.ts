import { apiClient } from "@/lib/api";
import { ChatInterface } from "@/types/chat";
import { requestHandler } from "@/utils/requestHandler";
import { useMutation } from "@tanstack/react-query";

export const useCreateNewChat = () => {
  return useMutation({
    mutationFn: (selectedUserId: string) =>
      requestHandler<ChatInterface>(() =>
        apiClient.post(`/chat/c/${selectedUserId}`),
      )(),
  });
};
