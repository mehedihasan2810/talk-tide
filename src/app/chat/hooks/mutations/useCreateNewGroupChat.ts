import { apiClient } from "@/lib/api";
import { ChatInterface } from "@/types/chat";
import { requestHandler } from "@/utils/requestHandler";
import { useMutation } from "@tanstack/react-query";

export const useCreateNewGroupChat = () => {
  return useMutation({
    mutationFn: ({
      groupName,
      participantIds,
    }: {
      groupName: string;
      participantIds: string[];
    }) =>
      requestHandler<ChatInterface>(() =>
        apiClient.post("/chatApp/group/create", {
          name: groupName,
          participantIds,
        }),
      )(),
  });
};
