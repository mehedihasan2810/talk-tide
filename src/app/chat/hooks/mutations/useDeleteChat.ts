import { queryClient } from "@/contexts/providers";
import { apiClient } from "@/lib/api";
import { SuccessResponse } from "@/types/api";
import { ChatInterface } from "@/types/chat";
import { requestHandler } from "@/utils/requestHandler";
import { useMutation } from "@tanstack/react-query";

export const useDeleteChat = () => {
  return useMutation({
    mutationFn: (chatId: string) =>
      requestHandler<ChatInterface>(() =>
        apiClient.delete(`/chatApp/chat/delete/${chatId}`),
      )(),

    onMutate: async (chatId: string) => {
      // Cancel any outgoing refetches
      // (so they don't overwrite our optimistic update)
      await queryClient.cancelQueries({ queryKey: ["chats"] });

      // Snapshot the previous value
      const prevChatsRes = queryClient.getQueryData([
        "chats",
      ]) as SuccessResponse<ChatInterface[]>;

      // Optimistically update to the new value
      queryClient.setQueryData(
        ["chats"],
        (oldChatsRes: SuccessResponse<ChatInterface[]>) => ({
          ...oldChatsRes,
          data: oldChatsRes.data.filter((chat) => chat.id !== chatId),
        }),
      );

      // Return a context object with the snapshotted value
      return {
        ...prevChatsRes,
        data: [...prevChatsRes.data],
      } as SuccessResponse<ChatInterface[]>;
    },

    // If the mutation fails,
    // use the context returned from onMutate to roll back
    onError: (_err, _newTodo, prevChatsRes) => {
      queryClient.setQueryData(["chats"], prevChatsRes);
    },

    // Always refetch after error or success:
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ["chats"] });
    },
  });
};
