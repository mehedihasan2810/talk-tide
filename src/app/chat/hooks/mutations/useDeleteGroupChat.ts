import { apiClient } from "@/lib/api";
import { ChatInterface } from "@/types/chat";
import { requestHandler } from "@/utils/requestHandler";
import { useMutation } from "@tanstack/react-query";

export const useDeleteGroupChat = () => {
  return useMutation({
    mutationFn: (chatId: string) =>
      requestHandler<ChatInterface>(() =>
        apiClient.delete(`/chatApp/group/delete/${chatId}`),
      )(),
  });
};
