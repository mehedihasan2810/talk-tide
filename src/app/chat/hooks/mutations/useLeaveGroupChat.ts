import { apiClient } from "@/lib/api";
import { ChatInterface } from "@/types/chat";
import { requestHandler } from "@/utils/requestHandler";
import { useMutation } from "@tanstack/react-query";

export const useLeaveGroupChat = () => {
  return useMutation({
    mutationFn: (chatId: string) =>
      requestHandler<ChatInterface>(() =>
        apiClient.patch(`/chatApp/group/leave/${chatId}`),
      )(),
  });
};
