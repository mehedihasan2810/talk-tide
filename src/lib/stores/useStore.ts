import { create } from "zustand";

interface ChatStore {
  isMobileSidebarOpen: boolean;
  isGroupDetailsModalOpen: boolean;

  toggleIsMobileSidebarOpen(): void;
  toggleGroupDetailsModal(_isGroupDetailsModalOpen: boolean): void;
}

export const useStore = create<ChatStore>((set) => ({
  // STATE
  isMobileSidebarOpen: false,
  isGroupDetailsModalOpen: false,
  // --------------------------

  // ACTION
  toggleIsMobileSidebarOpen: () =>
    set(({ isMobileSidebarOpen }) => ({
      isMobileSidebarOpen: !isMobileSidebarOpen,
    })),

  toggleGroupDetailsModal: (isGroupDetailsModalOpen) =>
    set({ isGroupDetailsModalOpen }),
}));
