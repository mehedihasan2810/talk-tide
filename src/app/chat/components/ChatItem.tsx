import { cn } from "@/lib/utils";
import { ChatInterface } from "@/types/chat";
import { SessionUser } from "@/types/session";
import { getChatObjectMetadata } from "@/utils/getChatObjectMetadata";
import {
  EllipsisVerticalIcon,
  InformationCircleIcon,
  PaperClipIcon,
  TrashIcon,
} from "@heroicons/react/20/solid";
import Image from "next/image";
import { FC } from "react";
import moment from "moment";
import { useStore } from "@/lib/stores/useStore";
import { usePathname, useRouter } from "next/navigation";
import { useDeleteChat } from "../hooks/mutations/useDeleteChat";
import { useCreateQueryString } from "../hooks/useCreateQueryString";
import {
  Popover,
  PopoverContent,
  PopoverTrigger,
} from "@/components/ui/popover";
import { Button, buttonVariants } from "@/components/ui/button";
import { Badge } from "@/components/ui/badge";
import {
  Dialog,
  DialogContent,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
  DialogFooter,
} from "@/components/ui/dialog";
import { useToast } from "@/components/ui/use-toast";

interface Props {
  chat: ChatInterface;
  onClick: (_chat: ChatInterface) => void;
  isActive?: boolean;
  unreadCount?: number;
  onChatDelete: (_chatId: string) => void;
  user: SessionUser;
}

const ChatItem: FC<Props> = ({
  chat,
  onClick,
  user,
  isActive,
  unreadCount = 0,
  onChatDelete,
}) => {
  const { toggleGroupDetailsModal } = useStore();

  const { mutate: deleteChatMutation } = useDeleteChat();

  const router = useRouter();
  const pathname = usePathname();
  const { toast } = useToast();

  // delete a specific query then return the remaining
  const createQueryString = useCreateQueryString();

  // if there is no chat then we don't want to render anything
  if (!chat) return;

  return (
    <div
      role="button"
      onClick={() => onClick(chat)}
      className={cn(
        "group relative isolate my-2 flex cursor-pointer items-start justify-between gap-3 overflow-hidden rounded-3xl border p-4 hover:bg-primary/5",
        isActive ? "border-primary  bg-primary/5" : "",
        unreadCount > 0 ? "border-primary bg-primary/10 font-bold" : "",
      )}
    >
      <div className="flex flex-shrink-0 items-center justify-center">
        {chat.isGroupChat ? (
          <div className="relative flex h-12 w-12 flex-shrink-0 flex-nowrap items-center justify-start">
            {chat.participants.slice(0, 3).map((participant, i) => {
              return (
                <Image
                  key={participant.id}
                  src={participant.avatar.url}
                  width={48}
                  height={48}
                  className={cn(
                    "outline-dark absolute h-7 w-7 rounded-full border-[1px] border-white outline outline-4 group-hover:outline-secondary",
                    i === 0
                      ? "left-0 z-[3]"
                      : i === 1
                      ? "left-2.5 z-[2]"
                      : i === 2
                      ? "left-[18px] z-[1]"
                      : "",
                  )}
                  alt="chat participant image"
                />
              );
            })}
          </div>
        ) : (
          <Image
            src={getChatObjectMetadata(chat, user).avatar as string}
            width={48}
            height={48}
            className="h-12 w-12 rounded-full"
            alt="chat participant image"
          />
        )}
      </div>

      <div className="w-full">
        <div className="flex gap-2">
          <p className="max-w-[200px] truncate">
            {" "}
            {getChatObjectMetadata(chat, user).title}
          </p>
          {/* Unread count will be > 0 when user is on another chat and there is new message in a chat which is not currently active on user's screen */}
          {unreadCount <= 0 ? null : (
            <Badge className="bg-primary/90">
              {unreadCount > 9 ? "9+" : unreadCount}
            </Badge>
          )}
        </div>
        <div className="inline-flex w-full items-center text-left">
          {chat.chatMessages[0] &&
          chat.chatMessages[0].attachments.length > 0 ? (
            // If last message is an attachment show paperclip
            <PaperClipIcon className="mr-2 flex h-3 w-3 flex-shrink-0 text-black/50" />
          ) : null}
          <small className="max-w-[150px] truncate text-sm text-zinc-500">
            {getChatObjectMetadata(chat, user).lastMessage}
          </small>

          <small className="ml-2 w-max font-semibold">
            {moment(chat.chatMessages[0]?.updatedAt || chat.updatedAt)
              .add("TIME_ZONE", "hours")
              .fromNow(true)}
          </small>
        </div>
      </div>

      <Popover>
        <PopoverTrigger
          onClick={(e) => {
            e.stopPropagation();
          }}
          className="absolute right-4 top-1/2 z-20 h-9 w-9 -translate-y-1/2 rounded-full border border-primary/30  p-1 text-primary group-hover:opacity-100 md:opacity-0"
        >
          <EllipsisVerticalIcon className=" h-6 w-6" />
        </PopoverTrigger>
        <PopoverContent
          side="top"
          className="w-fit border-none bg-none p-0 shadow-none"
        >
          {chat.isGroupChat ? (
            <Button
              onClick={(e) => {
                e.stopPropagation();

                router.replace(pathname + createQueryString("cD", chat.id), {
                  scroll: false,
                });

                toggleGroupDetailsModal(true);
              }}
              className="inline-flex items-center rounded-lg p-4"
            >
              <InformationCircleIcon className="mr-2 h-4 w-4" /> About group
            </Button>
          ) : (
            <Dialog>
              <DialogTrigger
                className={buttonVariants({
                  variant: "destructive",
                  className: "inline-flex items-center",
                  size: "sm",
                })}
                onClick={(e) => e.stopPropagation()}
              >
                <TrashIcon className="mr-2 h-4 w-4" />
                Delete chat
              </DialogTrigger>
              <DialogContent onClick={(e) => e.stopPropagation()}>
                <DialogHeader>
                  <DialogTitle className="text-center">
                    Are you sure you want to delete this chat?
                  </DialogTitle>
                </DialogHeader>

                <DialogFooter className="mx-auto">
                  <Button
                    onClick={(e) => {
                      e.stopPropagation();
                      deleteChatMutation(chat.id, {
                        onError: (error) => {
                          toast({
                           title: error.message,
                            variant: "destructive",
                          });
                        },
                      });
                      onChatDelete(chat.id);
                      toast({
                        description: "Chat deleted successfully",
                        variant: "success",
                      });
                    }}
                    size="sm"
                    variant="destructive"
                  >
                    <TrashIcon className="mr-2 h-4 w-4" />
                    Delete
                  </Button>
                </DialogFooter>
              </DialogContent>
            </Dialog>
          )}
        </PopoverContent>
      </Popover>
    </div>
  );
};

export default ChatItem;
