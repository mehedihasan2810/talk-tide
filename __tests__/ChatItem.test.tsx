import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import ChatItem from "@/app/chat/components/ChatItem";
import { useDeleteChat } from "@/app/chat/hooks/mutations/useDeleteChat";
import { useToast } from "@/components/ui/use-toast";
import { useRouter } from "next/navigation";
import { useStore } from "@/lib/stores/useStore";

// Mocking dependencies
jest.mock("@/lib/stores/useStore", () => ({
  useStore: jest.fn().mockReturnValue({
    toggleGroupDetailsModal: jest.fn(),
  }),
}));

jest.mock("@/app/chat/hooks/mutations/useDeleteChat", () => ({
  useDeleteChat: jest.fn(() => ({
    mutate: jest.fn(),
  })),
}));

jest.mock("@/app/chat/hooks/useCreateQueryString", () => ({
  useCreateQueryString: jest.fn(() => jest.fn()),
}));

jest.mock("@/components/ui/use-toast", () => ({
  useToast: jest.fn().mockReturnValue({
    toast: jest.fn(),
  }),
}));

// Mocking moment.js to keep the same time for testing
jest.mock("moment", () => {
  const momentMock = jest.requireActual("moment");
  return (date: string) => momentMock(date).utcOffset(0); // Mocking with UTC offset 0
});

jest.mock("next/navigation", () => ({
  usePathname: jest.fn(),
  useRouter: jest.fn().mockReturnValue({
    refresh: jest.fn(),
    replace: jest.fn(),
  }),
}));

jest.mock("@tanstack/react-query", () => ({
  useMutation: jest.fn(),
}));

// Dummy data for testing
const dummyUser = {
  id: "user1",
  avatar: {
    url: "avatar1.jpg",
    localPath: "local/avatar1.jpg",
  },
  username: "testuser",
  email: "test@example.com",
  isEmailVerified: true,
  loginType: "email",
  role: "user",
  createdAt: "2022-01-01T12:00:00Z",
  updatedAt: "2022-01-01T12:00:00Z",
};

const dummyChat = {
  id: "chat1",
  adminId: "admin1",
  isGroupChat: false,
  chatMessages: [
    {
      id: "message1",
      sender: dummyUser,
      senderId: dummyUser.id,
      content: "Hello!",
      chatId: "chat1",
      attachments: [],
      createdAt: "2022-01-01T12:30:00Z",
      updatedAt: "2022-01-01T12:30:00Z",
    },
  ],
  name: "Chat 1",
  participantIds: [dummyUser.id],
  participants: [dummyUser],
  createdAt: "2022-01-01T12:00:00Z",
  updatedAt: "2022-01-01T12:30:00Z",
};

const dummySessionUser = {
  id: "user1",
  email: "test@example.com",
  username: "testuser",
  role: "USER",
};

describe("ChatItem Component", () => {
  test("should render properly & match snapshot", () => {
    // ARRANGE
    const { asFragment } = render(
      <ChatItem
        chat={dummyChat}
        onClick={() => {}}
        user={dummySessionUser}
        onChatDelete={() => {}}
      />,
    );

    // ASSERT
    expect(asFragment()).toMatchSnapshot();
  });

  // --------------------------------------------------------------

  test("should handle click event", async () => {
    // MOCK
    const onClickMock = jest.fn();

    // EVENT
    const user = userEvent.setup();

    // ARRANGE
    render(
      <ChatItem
        chat={dummyChat}
        onClick={onClickMock}
        user={dummySessionUser}
        onChatDelete={() => {}}
      />,
    );

    // ACT
    await user.click(screen.getByTestId("chatItem-container"));

    // ASSERT
    expect(onClickMock).toHaveBeenCalledWith(dummyChat);
  });

  // --------------------------------------------------------------

  test("should render group chat avatars properly", async () => {
    // MOCK
    const groupChat = {
      ...dummyChat,
      isGroupChat: true,
      participantIds: ["user1", "user2", "user3"],
      participants: [
        { ...dummyUser, id: "user1" },
        { ...dummyUser, id: "user2" },
        { ...dummyUser, id: "user3" },
      ],
    };

    // EVENT
    const user = userEvent.setup();

    // ARRANGE
    render(
      <ChatItem
        chat={groupChat}
        onClick={() => {}}
        user={dummySessionUser}
        onChatDelete={() => {}}
      />,
    );

    // ACT
    await user.click(screen.getByTestId("pop-trigger"));
    await user.click(screen.getByRole("button", { name: "About group" }));

    // ASSERT
    expect(useRouter().replace).toHaveBeenCalled();
    expect(useStore().toggleGroupDetailsModal).toHaveBeenCalledWith(true);
    expect(screen.getAllByTestId("grp-part-avatar")).toHaveLength(3);
  });

  // ---------------------------------------------------------------

  test("should display the unread count badge", () => {
    const unreadCount = 5;
    render(
      <ChatItem
        chat={dummyChat}
        onClick={() => {}}
        user={dummySessionUser}
        onChatDelete={() => {}}
        unreadCount={unreadCount}
      />,
    );
    expect(screen.getByText(`${unreadCount}`)).toBeInTheDocument();
  });

  // --------------------------------------------------------

  test("should display '9+' for unread count greater than 9", () => {
    // MOCK
    const unreadCount = 15;

    // ARRANGE
    render(
      <ChatItem
        chat={dummyChat}
        onClick={() => {}}
        user={dummySessionUser}
        onChatDelete={() => {}}
        unreadCount={unreadCount}
      />,
    );

    // ASSERT
    expect(screen.getByText("9+")).toBeInTheDocument();
  });

  // --------------------------------------------------------

  test("should open the confirmation dialog on delete button click", async () => {
    // MOCK
    const onChatDeleteMock = jest.fn();

    // EVENT
    const user = userEvent.setup();

    // ARRANGE
    render(
      <ChatItem
        chat={dummyChat}
        onClick={() => {}}
        user={dummySessionUser}
        onChatDelete={onChatDeleteMock}
      />,
    );

    // ACT
    await user.click(screen.getByTestId("pop-trigger"));
    await user.click(screen.getByRole("button", { name: "Delete chat" }));

    // ASSERT
    expect(
      screen.getByText("Are you sure you want to delete this chat?"),
    ).toBeInTheDocument();
  });

  // -----------------------------------------------------------------------

  test("should close confirmation dialog on cancel button click", async () => {
    // EVENT
    const user = userEvent.setup();

    // ARRANGE
    render(
      <ChatItem
        chat={dummyChat}
        onClick={() => {}}
        user={dummySessionUser}
        onChatDelete={() => {}}
      />,
    );

    // ACT
    await user.click(screen.getByTestId("pop-trigger"));
    await user.click(screen.getByRole("button", { name: "Delete chat" }));
    await user.click(screen.getByRole("button", { name: "Close" }));

    // ASSERT
    expect(
      screen.queryByText("Are you sure you want to delete this chat?"),
    ).toBeNull();
  });

  // ------------------------------------------------------------------------------------

  test("should delete the chat on confirmation dialog delete button click", async () => {
    // MOCK
    const onChatDeleteMock = jest.fn();
    const deleteChatMutationMock = jest.fn();
    (useDeleteChat as jest.Mock).mockReturnValueOnce({
      mutate: deleteChatMutationMock,
    });
    const { toast } = useToast();

    // EVENT
    const user = userEvent.setup();

    // ARRANGE
    render(
      <ChatItem
        chat={dummyChat}
        onClick={() => {}}
        user={dummySessionUser}
        onChatDelete={onChatDeleteMock}
      />,
    );

    // ACT
    await user.click(screen.getByTestId("pop-trigger"));
    await user.click(screen.getByRole("button", { name: "Delete chat" }));
    await user.click(screen.getByRole("button", { name: "Delete" }));

    // ASSERT
    expect(deleteChatMutationMock).toHaveBeenCalledWith(dummyChat.id, {
      onError: expect.any(Function),
    });
    expect(onChatDeleteMock).toHaveBeenCalledWith(dummyChat.id);
    expect(toast).toHaveBeenCalledWith({
      description: "Chat deleted successfully",
      variant: "success",
    });
  });
});
