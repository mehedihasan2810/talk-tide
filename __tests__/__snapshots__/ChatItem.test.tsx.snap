// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatItem Component should render properly & match snapshot 1`] = `
<DocumentFragment>
  <div
    class="group relative isolate my-2 flex cursor-pointer items-start justify-between gap-3 overflow-hidden rounded-3xl border p-4 hover:bg-primary/5"
    data-testid="chatItem-container"
    role="button"
  >
    <div
      class="flex flex-shrink-0 items-center justify-center"
    >
      <span
        class="relative flex shrink-0 overflow-hidden rounded-full h-12 w-12"
      >
        <span
          class="flex h-full w-full items-center justify-center rounded-full bg-zinc-200"
        />
      </span>
    </div>
    <div
      class="w-full"
    >
      <div
        class="flex gap-2"
      >
        <p
          class="max-w-[200px] truncate"
        >
           
        </p>
      </div>
      <div
        class="inline-flex w-full items-center text-left"
      >
        <small
          class="max-w-[150px] truncate text-sm text-zinc-500"
        >
          You: Hello!
        </small>
        <small
          class="ml-2 w-max font-semibold"
        >
          2 years
        </small>
      </div>
    </div>
    <button
      aria-controls="radix-:r0:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="absolute right-4 top-1/2 z-20 h-9 w-9 -translate-y-1/2 rounded-full border border-primary/30  p-1 text-primary group-hover:opacity-100 md:opacity-0"
      data-state="closed"
      data-testid="pop-trigger"
      type="button"
    >
      <svg
        aria-hidden="true"
        class=" h-6 w-6"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 3a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM10 8.5a1.5 1.5 0 110 3 1.5 1.5 0 010-3zM11.5 15.5a1.5 1.5 0 10-3 0 1.5 1.5 0 003 0z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
